File: src/main/java/org/uno/engine/objects/NumericCard.java
/*
 * Copyright (C) 2020  Fábio Furtado
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.uno.engine.objects;

import org.uno.enums.CardColour;
import org.uno.enums.CardType;


/**
 * Immutable abstraction of a numeric card
 *
 * @author Fábio Furtado
 */
public final class NumericCard implements Card, Numeric, Colourful {

    private final CardColour colour;
    private final int number;
    private final CardType type;

    /**
     * Creates a new instance.
     *
     * @param colour colour of the card
     * @param number number the card will hold
     */
    public NumericCard(CardColour colour, int number) {
        this.type = CardType.NUMERIC;
        this.colour = colour;
        this.number = number;
    }

    /**
     * @see Card#getType()
     */
    public CardType getType() {
        return this.type;
    }

    /**
     * @see Colourful#getColour()
     */
    @Override
    public CardColour getColour() {
        return colour;
    }

    /**
     * @see Numeric#getNumber()
     */
    @Override
    public int getNumber() {
        return number;
    }

    /**
     * To be equal to this numeric card, the other card needs to also be numeric,
     * to have the same colour and number or to be a reference to this exact same
     * object.
     * @see Card#equals(Card)
     */
    @Override
    public boolean equals(Card other) {
        if (this == other)
            return true;
        if (this.getType() != other.getType())
            return false;
        NumericCard numOther = (NumericCard) other;
        return
            this.colour == numOther.getColour() &&
            this.number == numOther.getNumber();
    }

    /**
     * @see Card#clone()
     */
    @Override
    public Card clone() {
        return new NumericCard(this.getColour(), this.getNumber());
    }
}

File: src/main/java/org/uno/engine/objects/NumericCard.kt
/*
 * Copyright (C) 2020  Fábio Furtado
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.uno.engine.objects

import org.uno.engine.CardColour
import org.uno.engine.CardType
import kotlin.IllegalArgumentException

/**
 * Immutable abstraction of a numeric card which implements internment.
 *
 * @author Fábio Furtado
 */
class NumericCard private constructor(_colour: CardColour, _number: Int) :
                                                      Card, Numeric, Colourful {

    companion object {
        private val pool = HashMap<NumericCard, NumericCard>()

        /**
         * Creates a new instance
         *
         * @throws IllegalArgumentException if the number is not valid. A valid
         * one must be no shorter than [Numeric.MIN_VALUE] and no
         * longer than [Numeric.MAX_VALUE].
         */
        @JvmStatic
        fun of(_colour: CardColour, _number: Int): NumericCard {

            // validate the number
            if (_number < Numeric.MIN_VALUE || _number > Numeric.MAX_VALUE)
                throw IllegalArgumentException("$_number is not a valid number." +
                        "A valid one must be no shorter than ${Numeric.MIN_VALUE}" +
                        "and no longer than ${Numeric.MAX_VALUE}")

            val candidate = NumericCard(_colour, _number)
            return if (pool.containsKey(candidate)) pool[candidate]!!
            else {
                pool[candidate] = candidate
                candidate
            }
        }
    }

    /**
     * @see Card.type
     */
    override val type = CardType.NUMERIC

    /**
     * @see Colourful.colour
     */
    override val colour = _colour

    /**
     * @see Numeric.number
     */
    override val number = _number

    /**
     * Returns an instance with identical characteristics to this one except
     * for the new number.
     */
    fun withNumber(_number: Int) = of(colour, _number)


    /**
     * Returns an instance with identical characteristics to this one except
     * for the new colour.
     */
    fun withColour(_colour: CardColour) = of(_colour, number)

    /**
     * To be equal to this numeric card, the other card needs to also be numeric,
     * to have the same colour and number or to be a reference to this exact same
     * object.
     * @see Card.equals
     */
    fun equals(other: Card): Boolean {
        if (this == other) return true
        if (type != other.type) return false
        val numOther = other as NumericCard
        return colour == numOther.colour &&
                number == numOther.number
    }

    override fun equals(other: Any?): Boolean {
        return when {
            this === other -> true
            other == null || this.javaClass.kotlin != other.javaClass.kotlin -> false
            this.colour == (other as NumericCard).colour && this.number == other.number -> true
            else -> false
        }
    }
}